### PR TITLE
Fix account validation logic of legacy key to be used as an roleKey

### PR DESCRIPTION
**Pull request by aidan-kwon**
## Proposed changes

`(a *AccountKeyLegacy) Validate` can be called when `AccountRoleBased` contains `AccountKeyLegacy` as a role key like `AccountRoleBased{RoleTransaction: AccountKeyLegacy}`, but it always returns `false`.
 This change completes the method for `AccountKeyLegacy` to validate public keys recovered from tx signatures. 

- Complete `(a *AccountKeyLegacy) Validate` for the case when AccountKeyLegacy is used as a roleKey of RoleBasedKey.
- Rename a parameter of `Validate`: `pubkeys` to `recoveredKeys`
- Add a test case `TestAccountUpdateRoleBasedLegacy`

## Types of changes

Please put an x in the boxes related to your change.

- [x] Bugfix
- [ ] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: blockchain/types/accountkey/account_key.go [hunk 1]
```diff
@@ -108,16 +109,16 @@ func NewAccountKey(t AccountKeyType) (AccountKey, error) {
 	return nil, errUndefinedAccountKeyType
 }
 
-func ValidateAccountKey(from common.Address, accKey AccountKey, pubkeys []*ecdsa.PublicKey, roleType RoleType) error {
+func ValidateAccountKey(from common.Address, accKey AccountKey, recoveredKey []*ecdsa.PublicKey, roleType RoleType) error {
```
Comment (ehnuje) [inline blockchain/types/accountkey/account_key.go:112]:
```suggestion
func ValidateAccountKey(from common.Address, accKey AccountKey, recoveredKeys []*ecdsa.PublicKey, roleType RoleType) error {
```

File: blockchain/types/accountkey/account_key_weighted_multi_sig.go [hunk 2]
```diff
@@ -73,16 +75,16 @@ func (a *AccountKeyWeightedMultiSig) Equal(b AccountKey) bool {
...
 
 	// To prohibit making a signature with the same key, make a map.
 	// TODO-Klaytn: find another way for better performance
 	pMap := make(map[string]*ecdsa.PublicKey)
-	for _, bk := range pubkeys {
+	for _, bk := range recoveredKeys {
 		b, err := rlp.EncodeToBytes((*PublicKeySerializable)(bk))
 		if err != nil {
-			logger.Warn("Failed to encode public keys in the tx", pubkeys)
+			logger.Warn("Failed to encode public keys in the tx", recoveredKeys)
```
Comment (ehnuje) [inline blockchain/types/accountkey/account_key_weighted_multi_sig.go:87]:
```suggestion
			logger.Warn("Failed to encode public keys in the tx", "pubkeys", recoveredKeys)
```

File: tests/role_based_account_test.go [hunk 3]
```diff
@@ -407,6 +408,173 @@ func TestAccountUpdateRoleBasedNil(t *testing.T) {
 	}
 }
 
+// TestAccountUpdateRoleBasedLegacy tests TxInternalDataAccountUpdate with the following scenario:
+// 1. Update key to RoleBasedKey having a LegacyKey for all roles
+// 2. Test RoleTransfer of the RoleBasedKey
+// 3-1. Test RoleUpdate of the RoleBasedKey
+// 3-2. Recover the updated account key to the previously used RoleBasedKey
+// 4. Test RoleFeePayer of the RoleBasedKey
+func TestAccountUpdateRoleBasedLegacy(t *testing.T) {
```
Comment (ehnuje) [inline tests/role_based_account_test.go:417]: Does this test check the both `true` and `false` cases from `AccountKeyLegacy.Validate`?

Comment (aidan-kwon) [reply tests/role_based_account_test.go:419]: I added two more cases checking `false` return.

File: tests/role_based_account_test.go [hunk 4]
```diff
@@ -407,6 +408,222 @@ func TestAccountUpdateRoleBasedNil(t *testing.T) {
...
+
+	if testing.Verbose() {
+		fmt.Println("reservoirAddr = ", reservoir.Addr.String())
+		fmt.Println("anonAddr = ", anon.Addr.String())
+	}
+
+	signer := types.NewEIP155Signer(bcdata.bc.Config().ChainID)
+	gasPrice := new(big.Int).SetUint64(bcdata.bc.Config().UnitPrice)
+
+	// RoleBasedKey having LegacyKes for all roles
```
Comment (jeongkyun-oh) [inline tests/role_based_account_test.go:460]:
```suggestion
	// RoleBasedKey having LegacyKeys for all roles
```

File: blockchain/types/accountkey/account_key.go [hunk 5]
```diff
@@ -108,16 +109,16 @@ func NewAccountKey(t AccountKeyType) (AccountKey, error) {
 	return nil, errUndefinedAccountKeyType
 }
 
-func ValidateAccountKey(from common.Address, accKey AccountKey, pubkeys []*ecdsa.PublicKey, roleType RoleType) error {
+func ValidateAccountKey(from common.Address, accKey AccountKey, recoveredKeys []*ecdsa.PublicKey, roleType RoleType) error {
 	// Special treatment for AccountKeyLegacy.
 	if accKey.Type().IsLegacyAccountKey() {
```
Comment (kjhman21) [inline blockchain/types/accountkey/account_key.go:114]: @aidan-kwon If we implement validation logic in AccountKeyLegacy, it seems that this if statement is not necessary.

Comment (aidan-kwon) [reply blockchain/types/accountkey/account_key.go]: You're right. I removed it. Thanks!

File: tests/role_based_account_test.go (hunk 4 above)
Comment (aidan-kwon) [reply tests/role_based_account_test.go]: Changed. Thanks!

**Review by kjhman21 — APPROVED** [review]
LGTM. Thanks!

**Review by ehnuje — APPROVED** [review]

**Review by jeongkyun-oh — APPROVED** [review]

**Review by KimKyungup — APPROVED** [review]

**Pull request merged**